src/storage/src/streaming_table/state_table.rs

```diff
@@ -0,0 +1,426 @@
+        }
+    }
+}
+
+/// point get
+impl<S: StateStore> StateTable<S> {
+    /// Get a single row from state table.
+    pub async fn get_row<'a>(&'a self, pk: &'a Row, epoch: u64) -> StorageResult<Option<Row>> {
+        let serialized_pk = self.serialize_pk_with_vnode(pk);
+        let mem_table_res = self.mem_table.get_row_op(&serialized_pk);
+
+        let read_options = self.get_read_option(epoch);
+        match mem_table_res {
+            Some(row_op) => match row_op {
+                RowOp::Insert(row_bytes) => {
+                    let row = deserialize(self.mapping.clone(), serialized_pk, row_bytes)
+                        .map_err(err)?
+                        .2;
+                    Ok(Some(row))
+                }
+                RowOp::Delete(_) => Ok(None),
+                RowOp::Update((_, row_bytes)) => {
+                    let row = deserialize(self.mapping.clone(), serialized_pk, row_bytes)
+                        .map_err(err)?
+                        .2;
+                    Ok(Some(row))
+                }
+            },
+            None => {
+                if let Some(storage_row_bytes) =
+                    self.keyspace.get(&serialized_pk, read_options).await?
+                {
+                    let row = deserialize(self.mapping.clone(), &serialized_pk, &storage_row_bytes)
+                        .map_err(err)?
+                        .2;
+                    Ok(Some(row))
+                } else {
+                    Ok(None)
+                }
+            }
+        }
+    }
+
+    /// `vnode | pk`
+    fn serialize_pk_with_vnode(&self, pk: &Row) -> Vec<u8> {
+        let mut output = Vec::new();
+        output.put_slice(&self.compute_vnode_by_pk(pk).to_be_bytes());
+        self.pk_serializer.serialize(pk, &mut output);
+        output
+    }
+}
+
+// write
+impl<S: StateStore> StateTable<S> {
+    /// Insert a row into state table. Must provide a full row corresponding to the column desc of
+    /// the table.
+    pub fn insert(&mut self, value: Row) -> StorageResult<()> {
+        let mut datums = vec![];
+        for pk_index in self.pk_indices() {
+            datums.push(value.index(*pk_index).clone());
+        }
+        let pk = Row::new(datums);
+        let pk_bytes = serialize_pk(&pk, self.pk_serializer());
+        let vnode = self.compute_vnode_by_row(&value);
+        let (key_bytes, value_bytes) = serialize(vnode, &pk_bytes, value).map_err(err)?;
+        self.mem_table.insert(key_bytes, value_bytes);
+        Ok(())
+    }
+
+    /// Delete a row from state table. Must provide a full row of old value corresponding to the
+    /// column desc of the table.
+    pub fn delete(&mut self, old_value: Row) -> StorageResult<()> {
+        let mut datums = vec![];
+        for pk_index in self.pk_indices() {
+            datums.push(old_value.index(*pk_index).clone());
+        }
+        let pk = Row::new(datums);
+        let pk_bytes = serialize_pk(&pk, self.pk_serializer());
```

Comment:
```suggestion
        let mut datums = vec![];
        for pk_index in self.pk_indices() {
            datums.push(old_value.index(*pk_index).clone());
        }
        let pk = Row::new(self.pk_indices().map(|pk_index| old_value.index(*pk_index).clone()).collect());
```
Also, we might need a method `StateTable::get_pk_from_row`